[Tabs] Fix tab indicator flies off issue

nero-angela opened the pull request:
## Description
I think the code below is not only wrong, but unnecessary.
https://github.com/flutter/flutter/blob/54ade883652a74d9df45c46fdc7f4df9f06dfb37/packages/flutter/lib/src/material/tabs.dart#L400-L404

The reason why I think it's unnecessary is the below code already covers the case when the animation is running.
https://github.com/flutter/flutter/blob/54ade883652a74d9df45c46fdc7f4df9f06dfb37/packages/flutter/lib/src/material/tabs.dart#L406-L421

And as you can see the below code, it is implemented by putting `_currentRect` in `Rect.lerp()` and putting the result back in `_currentRect`. Implemented in this way, the indicator will performs an acceleration motion. I prepared a [dart pad](https://dartpad.dev/a8f99d990a069ebdd8c291459713853f) to help you understand.

https://github.com/flutter/flutter/blob/e682ec711a8c05fe78157595240b392de339fa32/packages/flutter/lib/src/material/tabs.dart#L403

In the test, I think below comment is mentioned because the indicator performs acceleration motion.
https://github.com/flutter/flutter/blob/e682ec711a8c05fe78157595240b392de339fa32/packages/flutter/test/material/tabs_test.dart#L1656

I have also identified the cause of the disappearance of the indicator, but I will skip this because the problematic code is replaced with the existing code.

|before|after|
|:--:|:--:|
|![tabs_before](https://user-images.githubusercontent.com/26322627/92576161-223a9d80-f2c4-11ea-8509-edc990232cae.gif)|![tabs_after](https://user-images.githubusercontent.com/26322627/92576122-1949cc00-f2c4-11ea-98a8-02303f546555.gif)|

## Related Issues

Fixes https://github.com/flutter/flutter/issues/48000

## Tests

I added the following test:
- Tab indicator animation test

I fixed the following test:
https://github.com/flutter/flutter/blob/e682ec711a8c05fe78157595240b392de339fa32/packages/flutter/test/material/tabs_test.dart#L1656-L1661

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

nero-angela committed:
[Tabs] Fix tab indicator flies off issue

nero-angela committed:
[Tabs] Add indicator animation test

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -397,29 +397,23 @@ class _IndicatorPainter extends CustomPainter {
+    else if (value == index)
+      _currentRect = middle;
+    else if (value < index)
+      _currentRect = previous == null ? middle : Rect.lerp(middle, previous, index - value);
```
isn't this assume index - value is always <1.0 ? in case of user tapp on a tab, this value may be >1.0

nero-angela replied:
Yes you are right. `index - value` can be greater than `1.0`. I thought this was the solution for the case of tapping a tab that more than one tab away. If there are any reason it's not allowed, please let me know.

I added a gif that simulated the situation you were talking about as I understood it. If this is not the case of you intended, could you describe the case of tapping a tab that more than one tab away for reproduce?

![tab_indicator_test](https://user-images.githubusercontent.com/26322627/93846955-86a42680-fce0-11ea-8512-b13df20f2751.gif)

nero-angela replied:
When there are three tabs A, B, C, the image below is a scenario in which the user taps C and then taps B after the animation has progressed 25%.

![tab](https://user-images.githubusercontent.com/26322627/98914552-2a08ff00-250c-11eb-89ad-99b2240687fc.png)

Here you can see a case which the lerp value is greater than 1, and I think it can be relatively greater or less than 1 depending on which point the lerp is based on.

As [this document](https://api.flutter.dev/flutter/dart-ui/Rect/lerp.html) also says that a value greater than 1 or less than 0 is not an invalid value, I think it is just a difference in mathematical expressions and there is no problem.

Nevertheless, if the lerp value should be between 0 and 1, I think that the function should be newly implemented.

chunhtai replied:
This seems to not working correctly if the sizes of the tab are not the same.
```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

void main() {
  timeDilation = 10.0;
  runApp(Test());
}

class Test extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: '',
      home: Material(
        color: Colors.black,
        child: DefaultTabController(
          length: 3,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text('Tap 1 then tap 2.', style: TextStyle(color: Colors.white)),
              TabBar(
                isScrollable: true,
                indicatorSize: TabBarIndicatorSize.tab,
                tabs: <Widget>[
                  Text('-', style: TextStyle(fontSize: 48.0)),
                  Text('1ljkjlkjlk', style: TextStyle(fontSize: 48.0)),
                  Text('2', style: TextStyle(fontSize: 48.0)),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

nero-angela replied:
You are right. It doesn't work properly. It seems my current approach is wrong.
Sorry for wasting your time. 🙏

chunhtai reviewed (commented):
I think we still need the first if case to deal with user tap a tab that are more than one tab away?

chunhtai commented:
Hi @nero-angela are you still working on this PR?

nero-angela commented:
> Hi @nero-angela are you still working on this PR?

Hi @chunhtai, I'm waitting for your answer about [this question](https://github.com/flutter/flutter/pull/65463#discussion_r492483111).

Could you please check my question?

nero-angela commented:
@chunhtai 
Going back to the beginning, I tried different approaches based on the current master branch. The code for the solution can be found [here](https://github.com/nero-angela/flutter/commit/38f2bc04ba84ae8d7e8a5986d0178157e5a6ad04).

At the moment of tabbing, the current indicator's position was recorded in the `TabController` as the `_indexPositionWhenTapping` property, and animation was created naturally by using it. I don't like the location of `_indexPositionWhenTapping`, but I couldn't find a better way to do this.

<img height="400px" src="https://user-images.githubusercontent.com/26322627/99063217-868d1c80-25e7-11eb-9e89-f0adc7943dc0.gif">

In this regard, is it better to close the current PR and create a new PR to proceed?

chunhtai commented:
I wonder whether we can do it in the following way:

we have `tabcontroller.value` and `tabcontroller.index`. with these information we know
1. The direction of the animation
2. The current tab index

We can try to grab the current tab index and the index that is one tab ahead. We can just lerp between this two tabs.

For example: value = 1.4,   index =5

we know the animation is moving to the right,
we can find the current tab index = math.floor(value) = 1
and the index one tab ahead = math.floor(value) + 1 = 2
we do lerp(tab1, tab2, value - math.floor(value))


for example  value = 2.3,   index = 1

we know the animation is moving to the left,
we can find the current tab index = math.ceil(value) = 3
and the index one tab ahead = math.ceil(value) -1  = 2
we do lerp(tab3, tab2, math.ceil(value) - value)


For abs(value - index) < 1.0 , we just keep the logic as is.

nero-angela committed:
[Tabs] Fix tab indicator flies off issue

nero-angela committed:
[Tabs] Fix the tabs test because the indicator's moving speed has been changed

nero-angela commented:
> I wonder whether we can do it in the following way:
> 
> we have `tabcontroller.value` and `tabcontroller.index`. with these information we know
> 
> 1. The direction of the animation
> 2. The current tab index
> 
> We can try to grab the current tab index and the index that is one tab ahead. We can just lerp between this two tabs.
> 
> For example: value = 1.4, index =5
> 
> we know the animation is moving to the right,
> we can find the current tab index = math.floor(value) = 1
> and the index one tab ahead = math.floor(value) + 1 = 2
> we do lerp(tab1, tab2, value - math.floor(value))
> 
> for example value = 2.3, index = 1
> 
> we know the animation is moving to the left,
> we can find the current tab index = math.ceil(value) = 3
> and the index one tab ahead = math.ceil(value) -1 = 2
> we do lerp(tab3, tab2, math.ceil(value) - value)
> 
> For abs(value - index) < 1.0 , we just keep the logic as is.

Wow, you are really smart 👍👍. I implemented the solution in the way you gave me. All of the cases found so far seem to work fine, but please check [the code](https://github.com/nero-angela/flutter/commit/4082e19387e7779b104b4c4aab0e885f1d9040bb) and let me know if it's not what you intended.

One of the previous tests failed because the indicator's speed was changed normally, and I fixed it correctly. Check out the test-related code [here](https://github.com/nero-angela/flutter/commit/a13448583f582311dff7eebedd41fd7ce4cc21eb).

chunhtai commented:
@nero-angela 
The code looks good. Now I think of it, we can probably apply the same logic when abs(value - index) < 1.0

nero-angela committed:
[Tabs] Delete unnecessary code

nero-angela commented:
@chunhtai 
I don't understand what state is when `abs(value-index) <1.0`.

I thought it would work well for all cases with the current code, so I deleted the existing unnecessary code.
[This code](https://github.com/nero-angela/flutter/commit/42854b328d3a7df967017c863d0ecd0d70941421) passed all the test code and it seems to work fine. Is this the code you intended?

chunhtai commented:
@nero-angela yes, I think that looks good, can you revise this PR with the change?

nero-angela committed:
Merge branch 'tabs_indicator' into tab_indicator

nero-angela committed:
Merge branch 'master' of https://github.com/flutter/flutter into tab_indicator

nero-angela commented:
@chunhtai 
I have finished revising the PR with the changes. Could you please check?

chunhtai reviewed (approved):
LGTM
